gtsfm/data_association/point3d_initializer.py

```diff
@@ -155,52 +171,58 @@ def triangulate(self, track_2d: SfmTrack2d) -> Tuple[Optional[SfmTrack], Optiona
             best_inliers = np.ones(len(track_2d.measurements), dtype=bool)  # all marked as inliers
 
+        # Verify we have at least 2 inliers.
         inlier_idxs = (np.where(best_inliers)[0]).tolist()
-
-        is_cheirality_failure = False
         if len(inlier_idxs) < 2:
-            return None, None, is_cheirality_failure
+            return None, None, TriangulationExitCode.INLIERS_UNDERCONSTRAINED
 
+        # Extract keypoint measurements corresponding to inlier indices.
         inlier_track = track_2d.select_subset(inlier_idxs)
+        track_cameras, track_measurements = self.extract_measurements(inlier_track)
+
+        # Exit if we do not have at least 2 measurements in cameras with estimated poses.
+        if track_cameras is None:
+            return None, None, TriangulationExitCode.POSES_UNDERCONSTRAINED
 
-        camera_track, measurement_track = self.extract_measurements(inlier_track)
+        # Triangulate and check for cheirality failure from GTSAM.
         try:
             triangulated_pt = gtsam.triangulatePoint3(
-                camera_track,
-                measurement_track,
+                track_cameras,
+                track_measurements,
                 rank_tol=SVD_DLT_RANK_TOL,
                 optimize=True,
             )
         except RuntimeError:
-            is_cheirality_failure = True
-            return None, None, is_cheirality_failure
+            return None, None, TriangulationExitCode.CHEIRALITY_FAILURE
 
-        # compute reprojection errors for each measurement
+        # Compute reprojection errors for each measurement.
         reproj_errors, avg_track_reproj_error = reproj_utils.compute_point_reprojection_errors(
             self.track_camera_dict, triangulated_pt, inlier_track.measurements
         )
 
-        # all the measurements should have error < threshold
+        # Check that all the measurements have reprojection error < threshold.
+        # TODO(johnwlambert): compare with approach where we only throw away the outlier measurements.
         if not np.all(reproj_errors < self.reproj_error_thresh):
-            return None, avg_track_reproj_error, is_cheirality_failure
+            return None, avg_track_reproj_error, TriangulationExitCode.EXCEEDS_REPROJ_THRESH
 
+        # Create a gtsam.SfmTrack with the triangulated 3d point and associated 2d measurements.
         track_3d = SfmTrack(triangulated_pt)
-        for i, uv in inlier_track.measurements:
-            track_3d.add_measurement(i, uv)
+        for camera_id, uv in inlier_track.measurements:
+            track_3d.add_measurement(camera_id, uv)
 
-        return track_3d, avg_track_reproj_error, is_cheirality_failure
+        return track_3d, avg_track_reproj_error, TriangulationExitCode.SUCCESS
 
-    def generate_measurement_pairs(self, track: SfmTrack2d) -> List[Tuple[int, int]]:
+    def generate_measurement_pairs(self, track: SfmTrack2d) -> List[Tuple[int, ...]]:
```

Comment:
why the `...`?

Reply:
`mypy` throws an error because the return type of `itertools.combinations` is `Tuple[int, ...]`. I think using this type makes more sense anyways since the actual length of the tuple is variable since it's specified by `NUM_SAMPLES_PER_RANSAC_HYPOTHESIS`.

Reply:
ah i see, cool